Source/CarthageKit/Project.swift

```diff
@@ -866,54 +858,56 @@ private func repositoryURLForProject(project: ProjectIdentifier, preferHTTPS: Bo
-			} catch let error as NSError {
-				return .Failure(.WriteFailed(CarthageDependencyRepositoriesURL, error))
-			}
-
-			return .Success(repositoryURLForProject(project, preferHTTPS: preferHTTPS))
+		do {
+			try fileManager.createDirectoryAtURL(CarthageDependencyRepositoriesURL, withIntermediateDirectories: true, attributes: nil)
+		} catch let error as NSError {
+			return .Failure(.WriteFailed(CarthageDependencyRepositoriesURL, error))
 		}
-		.flatMap(.Merge) { remoteURL -> SignalProducer<(ProjectEvent, NSURL), CarthageError> in
-			let cloneProducer: () -> SignalProducer<(ProjectEvent, NSURL), CarthageError> = {
-				let cloneProducer = cloneRepository(remoteURL, repositoryURL)
 
-				return SignalProducer(value: (.Cloning(project), repositoryURL))
-					.concat(cloneProducer.then(.empty))
-			}
-
-			if let _ = try? fileManager.createDirectoryAtURL(repositoryURL, withIntermediateDirectories: false, attributes: nil) {
-				// If we created the directory, we're now responsible for
-				// cloning it.
-				return cloneProducer()
-			} else {
-				return isGitRepository(repositoryURL)
-					.promoteErrors(CarthageError.self)
-					.flatMap(.Concat) { isRepository -> SignalProducer<(ProjectEvent, NSURL), CarthageError> in
-						if isRepository {
-							let fetchProducer = fetchRepository(repositoryURL, remoteURL: remoteURL, refspec: "+refs/heads/*:refs/heads/*") /* lol syntax highlighting */
-
-							return SignalProducer(value: (.Fetching(project), repositoryURL))
-								.concat(fetchProducer.then(.empty))
-						} else {
-							// If the directory isn't a repository (that might
-							// happen if the process is quitted right after
-							// creating the directory), remove the directory then
-							// clone it.
-							_ = try? fileManager.removeItemAtURL(repositoryURL)
-							return cloneProducer()
+		return .Success(repositoryURLForProject(project, preferHTTPS: preferHTTPS))
+	}
+	.flatMap(.Merge) { remoteURL -> SignalProducer<(ProjectEvent, NSURL), CarthageError> in
+		return isGitRepository(repositoryURL)
+			.promoteErrors(CarthageError.self)
+			.flatMap(.Merge) { isRepository -> SignalProducer<(ProjectEvent, NSURL), CarthageError> in
+				if isRepository {
+					let fetchProducer: () -> SignalProducer<(ProjectEvent, NSURL), CarthageError> = {
+						return SignalProducer(value: (.Fetching(project), repositoryURL))
+							.concat(fetchRepository(repositoryURL, remoteURL: remoteURL, refspec: "+refs/heads/*:refs/heads/*").then(.empty))
+					}
+					// If we've already cloned the repo, check for the revision, possibly skipping an unnecessary fetch
+					if let commitish = commitish {
+						return commitExistsInRepository(repositoryURL, revision: commitish)
+							.promoteErrors(CarthageError.self)
+							.flatMap(.Merge) { exists -> SignalProducer<(ProjectEvent, NSURL), CarthageError> in
+								if exists {
+									return SignalProducer(value: (.SkippedFetching(project), repositoryURL))
```

Comment:
Rather than adding a `SkippedFetching` event and then ignoring it, could we change the signature of this method to return `SignalProducer<(ProjectEvent?, NSURL), CarthageError>` and then return `nil` for the project event here? That seems like it'd make a bit more sense.